Add a post about changes to WebAssembly targets (#1385)
rust-lang/blog.rust-lang.org: posts/2024-08-26-webassembly-targets-change-in-default-target-features.md

like:

1. A new WebAssembly proposal is created in a new repository, for example
   WebAssembly/foo.
2. Eventually Rust/LLVM implement the proposal under `-Ctarget-feature=+foo`
3. Eventually the upstream proposal is merged into the specification, and
   WebAssembly/foo becomes an archived repository
4. Rust/LLVM enable the `-Ctarget-feature=+foo` feature by default but typically
   retain the ability to disable it as well.

The `reference-types` and `multivalue` target features in Rust are at step (4)
here now and this post is explaining the consequences of doing so.

## Enabling Reference Types by Default

The [reference-types proposal to
WebAssembly](https://github.com/webAssembly/reference-types) introduced a few
new concepts to WebAssembly, notably the `externref` type which is a
host-defined GC resource that WebAssembly cannot access but can pass around.
Rust does not have support for the WebAssembly `externref` type and LLVM 19 does
not change that. WebAssembly modules produced from Rust will continue to not use
the `externref` type nor have a means of being able to do so. This may be
enabled in the future, but it will mostly likely only be done on an opt-in basis

> and will not affect preexisting code by default.

Author: I mostly just put this here to be cautious. Bugs happen and people can be surprised to the point that they aren't aware something is a but, so I wanted to be sure to be clear here that the intention isn't that one day `externref` just sort of shows up in modules.

means that when an object file is emitted it does not know the final index of
the table that is going to be used in the final binary. Before reference-types
there was only one option, table 0, so `0x00` was always used when encoding
`call_indirect` instructions. After reference-types, however, LLVM will emit an
over-long [LEB] of the form `0x80 0x80 0x80 0x80 0x00` which is the maximal
length of a 32-bit [LEB]. This [LEB] is then filled in by the linker with a
relocation to the actual table index that is used by the final module.

When putting all of this together, it means that with LLVM 19, which has
the `reference-types` feature enabled by default, any WebAssembly module with an
indirect function call (which is almost always the case for Rust code) will
produce a WebAssembly binary that cannot be decoded by engines and tooling that
do not support the reference-types proposal. It is expected that this change
will have a low impact due to the age of the reference-types proposal and
breadth of implementation in engines. Given the multitude of WebAssembly
engines, however, it's recommended that any WebAssembly users test out

> Nightly Rust and see if the produced module still runs on the engine of

Reviewer: ```suggestion
Nightly Rust and see if the produced module still runs on their engine of
```

continues to match LLVM's. Despite this though the change has the possibility
of still affecting Nightly users of Rust.

Rust for some time has supported an `extern "wasm"` ABI on Nightly which was an
experimental means of exposing the ability of defining a function in Rust which
returned multiple values (e.g. used the multi-value proposal). Due to
infrastructural changes and refactorings in LLVM itself this feature of Rust has
[been removed](https://github.com/rust-lang/rust/pull/127605) and is no longer
supported on Nightly at all. As a result there is no longer any possible method
of writing a function in Rust that returns multiple values at the WebAssembly
function type level.

In summary this change is expected to not affect any Rust code in the wild
unless you were using the Nightly feature of `extern "wasm"` in which case
you'll be forced to drop support for that and use `extern "C"` instead.
Supporting WebAssembly multi-return functions in Rust is a broader topic than
this post can cover, but at this time it's an area that's ripe for contribution
from suitably motivated contributors.

### Aside: ABI Stability and WebAssembly

> While on the topics of ABIs and the `multi-value` feature it's perhaps worth

Reviewer: ```suggestion
While on the topics of ABIs and the `multivalue` feature it's perhaps worth
```

> While on the topics of ABIs and the `multi-value` feature it's perhaps worth

Reviewer: ```suggestion
While on the topic of ABIs and the `multi-value` feature it's perhaps worth
```
Not sure.

also going over a bit what ABIs mean for WebAssembly. Currently there's
effectively only one ABI implemented by LLVM and the Rust compiler meaning that
at the WebAssembly module level `extern "C"` and `extern "Rust"` are
more-or-less the same. Despite this though there's no need for this to be true.

The `extern "C"` ABI, what C code uses by default as well, is difficult to
change because stability is often required across different compiler versions.
For example WebAssembly code compiled with LLVM 18 might be expected to work
with code compiled by LLVM 20. This means that changing the ABI is a daunting
task that requires version fields, explicit markers, etc, to help prevent
mismatches.

> The `extern "Rust"` ABI, however, is entirely within the control of the Rust

Reviewer: We may want to mention here that `extern "Rust" is considered unstable anyway and this isn't a Wasm-specific issue.
However I was unable to find a specific quote for this in the Rust reference.

Reviewer 2: Rust ABI is perma-unstable AFAIK, there's the crABI initiative but that's not the "Rust" ABI.

compiler and can change from version to version. Just because it happens to work
like `extern "C"` today doesn't mean it will continue to do so in the future. A

> great example of this is that when the `multi-value` feature is enabled the

Reviewer: ```suggestion
great example of this is that when the `multivalue` feature is enabled the
```

  either be done on your engine's repository, the Rust repository, or the
  WebAssembly
  [tool-conventions](https://github.com/WebAssembly/tool-conventions)
  repository. It's recommended to first search to confirm there isn't already an
  open issue though.
* Recompile your code with features disabled, more on this in the next section.

The general assumption behind enabling new features by default is that it's a
relatively hassle-free operation for end users while bringing performance
benefits for everyone (e.g. nontrapping-fp-to-int will make float-to-int
conversions more optimal). If updates end up causing hassle it's best to flag
that early on so rollout plans can be adjusted if needed.

## Disabling on-by-default WebAssembly proposals

For a variety of reasons you might be motivated to disable on-by-default
WebAssembly features: for example maybe your engine is difficult to update or
doesn't support a new feature. Disabling on-by-default features is unfortunately
not the easiest task. It is notably not sufficient to use
`-Ctarget-features=-foo` to disable features for just your own project's
compilation because the Rust standard library, shipped in precompiled form, is

> compiled with this features enabled.

Reviewer: ```suggestion
compiled with these features enabled.
```

---
layout: post
title: "WebAssembly targets: change in default target-features"
author: Alex Crichton
team: The Compiler Team <https://www.rust-lang.org/governance/teams/compiler>
---

The Rust compiler has [recently upgraded to using LLVM 19][llvm19] and this
change accompanies some updates to the default set of target features enabled

> for WebAssembly targets of the Rust compiler. Nightly Rust today, which will

Reviewer 3: ```suggestion
for WebAssembly targets of the Rust compiler. Beta Rust today, which will
```

means that when an object file is emitted it does not know the final index of
the table that is going to be used in the final binary. Before reference-types
there was only one option, table 0, so `0x00` was always used when encoding
`call_indirect` instructions. After reference-types, however, LLVM will emit an
over-long [LEB] of the form `0x80 0x80 0x80 0x80 0x00` which is the maximal
length of a 32-bit [LEB]. This [LEB] is then filled in by the linker with a
relocation to the actual table index that is used by the final module.

When putting all of this together, it means that with LLVM 19, which has
the `reference-types` feature enabled by default, any WebAssembly module with an
indirect function call (which is almost always the case for Rust code) will
produce a WebAssembly binary that cannot be decoded by engines and tooling that
do not support the reference-types proposal. It is expected that this change
will have a low impact due to the age of the reference-types proposal and
breadth of implementation in engines. Given the multitude of WebAssembly
engines, however, it's recommended that any WebAssembly users test out

> Nightly Rust and see if the produced module still runs on their engine of

Reviewer 3: ```suggestion
Rust 1.82 beta and see if the produced module still runs on their engine of
```

typically optimize LLVM's default output even further. When optimizing a module
through [`wasm-opt`] these 5-byte encodings of index 0 are all shrunk to a
single byte.

## Enabling Multi-Value by Default

The second feature enabled by default in LLVM 19 is `multivalue`. The
[multi-value proposal to WebAssembly][multi-value] enables functions to have
more than one return value for example. WebAssembly instructions are
additionally allowed to have more than one return value as well. This proposal
is one of the first to get merged into the WebAssembly specification after the
original MVP and has been implemented in many engines for quite some time.

The consequences of enabling this feature by default in LLVM are more minor for
Rust, however, than enabling the `reference-types` feature by default. LLVM's
default C ABI for WebAssembly code is not changing even when `multivalue` is
enabled.  Additionally Rust's `extern "C"` ABI for WebAssembly is not changing
either and continues to match LLVM's (or strives to, [differences to
LLVM](https://github.com/rust-lang/rust/issues/115666) are considered bugs to
fix). Despite this though the change has the possibility of still affecting

> Nightly users of Rust.

Reviewer 3: ```suggestion 
Rust users.
```

WebAssembly feature that the `multivalue` enables, is still out on the horizon
and not implemented. First LLVM will need to implement complete lowering support
to generate WebAssembly functions with multiple returns, and then `extern
"Rust"` can be change to use this when fully supported. In the yet-further-still
future C code might be able to change, but that will take quite some time due to
its cross-version-compatibility story.

## Enabling Future Proposals to WebAssembly

This is not the first time that a WebAssembly proposal has gone from
off-by-default to on-by-default in LLVM, nor will it be the last. For example
LLVM already enables the [sign-extension proposal][sign-ext] by default which
MVP WebAssembly did not have. It's expected that in the not-too-distant future
the
[nontrapping-fp-to-int](https://github.com/WebAssembly/nontrapping-float-to-int-conversions)
proposal will likely be enabled by default. These changes are currently not made
with strict criteria in mind (e.g. N engines must have this implemented for M
years), and there may be breakage that happens.

If you're using a WebAssembly engine that does not support the modules emitted

> by Nightly Rust and LLVM 19 then your options are:

Reviewer 3: ```suggestion
by Rust 1.82 beta and LLVM 19 then your options are:
```

this post can cover, but at this time it's an area that's ripe for contribution
from suitably motivated contributors.

### Aside: ABI Stability and WebAssembly

While on the topics of ABIs and the `multi-value` feature it's perhaps worth
also going over a bit what ABIs mean for WebAssembly. Currently there's
effectively only one ABI implemented by LLVM and the Rust compiler meaning that

> at the WebAssembly module level `extern "C"` and `extern "Rust"` are

Reviewer 4: `extern "Rust"` doesn't have the struct unpacking behavior that `extern "C"` has on wasm32-unknown-unknown. And `extern "Rust"` passes small structs as up to two arguments, while `extern "C"` on wasm32-wasi and wasm32-unknown-emscripten would pass structs by-ref if they have more than a single primitive field.

Author: That's an excellent point and reminds me to be more precise here. I tend to always forget "ABI" additionally refers to passing structures by value which means my own communication about this isn't always precise as how you pass structures by value is rightfully a huge part of how ABIs work. I've reworded this section to remove saying that `extern "C"` is the same as `extern "Rust"` but still explain how `extern "C"` should match Clang and is hard to change while changing `extern "Rust"` should be possible with LLVM support.
